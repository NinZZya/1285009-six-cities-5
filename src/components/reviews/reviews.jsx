import React from 'react';
import Review from './components/review/review';
import * as Type from '../../constants/types';


const Reviews = ({reviews = []}) => {

  return (
    <ul className="reviews__list">
      {reviews.map((review) => (
        <Review key={`review-${review.id}`} review={review} />
      ))}
    </ul>
  );
};

Reviews.propTypes = {
  reviews: Type.REVIEWS,
};


export default Reviews;
