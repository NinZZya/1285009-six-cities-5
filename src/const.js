export const AppPath = {
  ROOT: `/`,
  LOGIN: `/login`,
  CITY: `/city`,
  OFFER: `/offer`,
  FAVORITES: `/favorites`,
  NOT_FOUND: `/not-found`,
};

export const CITIES = {
  1: {id: 1, name: `Paris`},
  2: {id: 2, name: `Cologne`},
  3: {id: 3, name: `Brussels`},
  4: {id: 4, name: `Amsterdam`},
  5: {id: 5, name: `Hamburg`},
  6: {id: 6, name: `Dusseldorf`},
};

export const SortType = {
  POPULAR: `Popular`,
  LOW_TO_HIGH: `Price: low to high`,
  HIGH_TO_LOW: `Price: high to low`,
  TOP_RATED_FIRST: `Top rated first`,
};

export const STARS = [
  {value: 5, title: `perfect`},
  {value: 4, title: `good`},
  {value: 3, title: `not bad`},
  {value: 2, title: `badly`},
  {value: 1, title: `terribly`},
];
