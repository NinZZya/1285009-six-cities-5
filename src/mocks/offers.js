export default [
  {
    id: 0,
    city: {
      id: 1,
      name: `Paris`,
    },
    title: `Wood and stone place`,
    type: `Private room`,
    price: 228,
    rate: 0,
    bedroomsCount: 1,
    adultsCount: 1,
    features: [
      `Dishwasher`,
      `Baby seat`,
    ],
    host: {
      name: `Angelina`,
      avatar: `/img/avatar-angelina.jpg`,
      isPro: false,
    },
    images: [
      `/img/apartment-03.jpg`,
      `/img/apartment-02.jpg`,
    ],
    description: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    reviews: [
      {
        id: 0,
        user: {
          id: 3,
          name: `John`,
          avatar: `/img/avatar.svg`,
        },
        date: `2019-11-26`,
        rate: 2,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 1,
        user: {
          id: 3,
          name: `John`,
          avatar: `/img/avatar.svg`
        },
        date: `2019-03-23`,
        rate: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      }
    ],
    isPremium: false,
  },
  {
    id: 1,
    city: {
      id: 1,
      name: `Paris`,
    },
    title: `Canal View Prinsengracht`,
    type: `Apartment`,
    price: 97,
    rate: 3,
    bedroomsCount: 1,
    adultsCount: 3,
    features: [
      `Dishwasher`,
      `Towels`,
      `Fridge`,
      `Heating`,
      `Washing machine`,
      `Wi-Fi`,
      `Coffee machine`,
    ],
    host: {
      name: `John`,
      avatar: `/img/avatar.svg`,
      isPro: false,
    },
    images: [
      `/img/apartment-01.jpg`,
      `/img/apartment-02.jpg`,
    ],
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    reviews: [
      {
        id: 2,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`
        },
        date: `2018-09-22`,
        rate: 2,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 3,
        user: {
          id: 1,
          name: `Max`,
          avatar: `/img/avatar-max.jpg`
        },
        date: `2018-10-24`,
        rate: 0,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 4,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`
        },
        date: `2019-08-16`,
        rate: 1,
        text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      }
    ],
    isPremium: false,
  },
  {
    id: 2,
    city: {
      id: 2,
      name: `Paris`
    },
    title: `Beautiful & luxurious apartment at great location`,
    type: `Apartment`,
    price: 89,
    rate: 1,
    bedroomsCount: 2,
    adultsCount: 1,
    features: [],
    host: {
      name: `Angelina`,
      avatar: `/img/avatar-angelina.jpg`,
      isPro: false,
    },
    images: [
      `/img/apartment-01.jpg`
    ],
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    reviews: [
      {
        id: 5,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`,
        },
        date: `2019-06-07`,
        rate: 3,
        text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      },
      {
        id: 6,
        user: {
          id: 1,
          name: `Max`,
          avatar: `/img/avatar-max.jpg`,
        },
        date: `2019-06-02`,
        rate: 1,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      }
    ],
    isPremium: false,
  },
  {
    id: 3,
    city: {
      id: 2,
      name: `Paris`,
    },
    title: `Wood and stone place`,
    type: `Private room`,
    price: 381,
    rate: 3,
    bedroomsCount: 1,
    adultsCount: 1,
    features: [
      `Fridge`,
      `Towels`,
    ],
    host: {
      name: `Angelina`,
      avatar: `/img/avatar-angelina.jpg`,
      isPro: false
    },
    images: [
      `/img/room.jpg`,
      `/img/apartment-02.jpg`,
      `/img/apartment-03.jpg`,
      `/img/studio-01.jpg`,
    ],
    description: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    reviews: [
      {
        id: 7,
        user: {
          id: 4,
          name: `Ann`,
          avatar: `/img/avatar.svg`,
        },
        date: `2018-11-08`,
        rate: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      },
      {
        id: 8,
        user: {
          id: 2,
          name: `Angelina`,
          avatar: `/img/avatar-angelina.jpg`
        },
        date: `2019-02-23`,
        rate: 0,
        text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      }
    ],
    isPremium: false,
  }
];
