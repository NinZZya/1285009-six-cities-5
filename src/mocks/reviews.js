export default [
  {
    id: 0,
    user: {
      id: 3,
      name: `John`,
      avatar: `/img/avatar.svg`,
      isPro: true,
    },
    date: new Date(`2019-11-26`),
    rate: 2,
    text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
  },
  {
    id: 1,
    user: {
      id: 3,
      name: `John`,
      avatar: `/img/avatar.svg`,
      isPro: false,
    },
    date: new Date(`2019-03-23`),
    rate: 4,
    text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
  }
];
