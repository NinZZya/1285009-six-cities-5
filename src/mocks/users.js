export default [{
  id: 0,
  name: `test`,
  email: `test@test.test`,
  password: `test`,
}];
