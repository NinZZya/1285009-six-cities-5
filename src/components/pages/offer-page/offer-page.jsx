import React, {useEffect} from 'react';
import {Redirect} from 'react-router-dom';
import {connect} from 'react-redux';
import PageContainer from '../../page-container/page-container';
import Container from '../../container/container';
import OfferGallery from '../../offer-gallery/offer-gallery';
import OfferMark from '../../offer-mark/offer-mark';
import BookmarkButton, {BookmarkButtonType} from '../../bookmark-button/bookmark-button';
import RaitingStars from '../../raiting-stars/raiting-stars';
import OfferFeatures from '../../offer-features/offer-features';
import OfferFeature from '../../offer-features/components/offer-feature/offer-feature';
import OfferPrice from '../../offer-price/offer-price';
import OfferInside from '../../offer-inside/offer-inside';
import OfferHost from '../../offer-host/offer-host';
import Reviews from '../../reviews/reviews';
import NewReview from '../../reviews/components/new-review/new-review';
import Map from '../../map/map';
import LoadingData from '../../loading-data/loading-data';
import OffersList, {OffersListType} from '../../offers-list/offers-list';
import * as UserSelector from '../../../store/reducer/user/user-selectors';
import * as OffersOperation from '../../../store/reducer/offers/offers-operations';
import * as OffersSelector from '../../../store/reducer/offers/offers-selectors';
import * as Type from '../../../constants/types';
import {AppPath, IdName, DataStatus, UserStatus} from '../../../constants/const';
import {extend} from '../../../utils/utils';


const NEAR_OFFERS_COUNT = 3;
const ContainerType = {
  PAGE: `property`,
  GALLERY: `property__gallery`,
};

const FetureType = {
  ENTIRE: `entire`,
  BEDROOMS: `bedrooms`,
  ADULTS: `adults`,
};

const TypeName = {
  OFFER_PRICE: `property`,
  RAITING_STARS: `property__stars`,
  MARK: `property`,
};

const OFFER_CONTAINER_CLASS_NAME = `property__container`;

const getOfferId = (match) => {
  const matchId = match.params[IdName.OFFER];
  const isCoorectId = matchId ? matchId.slice(0, 2) !== `00` : false;
  const offerId = Number(matchId);

  return !isNaN(offerId) && isCoorectId ? offerId : -1;
};

const renderReviewCount = (count) => {
  return (
    <>
      &middot; <span className="reviews__amount">{count}</span>
    </>
  );
};

const getReviewsContent = (args) => {
  const {
    userStatus,
    user,
    reviewsStatus,
    reviews,
    handelSubmitReview,
  } = args;

  const count = reviews.length;
  const isAuth = userStatus === UserStatus.AUTH && user;

  const reviewsContent = (
    reviewsStatus !== DataStatus.LOADING && reviewsStatus !== DataStatus.ERROR
  ) ?
    <Reviews reviews={reviews} /> :
    null;

  return (
    <section className="property__reviews reviews">
      <h2 className="reviews__title">
        Reviews {count ? renderReviewCount(count) : null}
      </h2>
      <LoadingData status={reviewsStatus} />
      {reviewsContent}
      {isAuth && (
        <NewReview
          reviewsStatus={reviewsStatus}
          onSubmitReview={handelSubmitReview}
        />
      )}
    </section>
  );
};

const getOfferContent = (args) => {
  const {
    offer,
    nearOffersStatus,
  } = args;

  const {
    images,
    isPremium,
    isFavorite,
    title,
    rate,
    type,
    bedroomsCount,
    adultsCount,
    features,
    price,
    host,
    description,
    coords,
    zoom,
  } = offer;

  const nearOffers = args.nearOffers.slice(0, NEAR_OFFERS_COUNT);
  const mapNearOffers = [...nearOffers, offer];
  const nearOffersContent = nearOffersStatus === DataStatus.SUCCESS ?
    <OffersList type={OffersListType.NEAR} offers={nearOffers} /> :
    null;

  const center = {
    coords,
    zoom,
  };

  return (
    <>
      <section className="property">
        <Container type={ContainerType.GALLERY}>
          <OfferGallery images={images} />
        </Container>
        <Container className={OFFER_CONTAINER_CLASS_NAME}>
          <div className="property__wrapper">
            {isPremium && <OfferMark type={TypeName.MARK} />}
            <div className="property__name-wrapper">
              <h1 className="property__name">
                {title}
              </h1>
              <BookmarkButton
                type={BookmarkButtonType.OFFER}
                mark={isFavorite}
                offer={offer}
              />
            </div>
            <div className="property__rating rating">
              <RaitingStars type={TypeName.RAITING_STARS} rate={rate} />
              <span className="property__rating-value rating__value">{rate}</span>
            </div>
            <OfferFeatures>
              <OfferFeature type={FetureType.ENTIRE} value={type} />
              <OfferFeature type={FetureType.BEDROOMS} value={`${bedroomsCount} Bedrooms`} />
              <OfferFeature type={FetureType.ADULTS} value={`Max ${adultsCount} adults`} />
            </OfferFeatures>
            <OfferPrice type={TypeName.OFFER_PRICE} price={price} />
            {features.length ? <OfferInside features={features} /> : null}
            <OfferHost host={host} description={description} />
            {getReviewsContent(args)}
          </div>
        </Container>
        <section className="property__map map">
          <Map
            center={center}
            pins={mapNearOffers}
            activeId={offer.id}
          />
        </section>
      </section>
      <Container>
        <section className="near-places places">
          <h2 className="near-places__title">
            Other places in the neighbourhood
          </h2>
          <LoadingData status={nearOffersStatus} />
          {nearOffersContent}
        </section>
      </Container>
    </>
  );
};

const OfferPage = (props) => {
  const {
    onLoadOffer,
    onLoadReviews,
    onLoadNearOffers,
    offerStatus,
    onGetOffer,
    onAddReview,
    match = {path: `/`},
  } = props;

  const activeOfferId = getOfferId(match);
  const offer = offerStatus === DataStatus.SUCCESS ?
    onGetOffer(activeOfferId) :
    null;

  if (activeOfferId === -1 || (!offer && offerStatus === DataStatus.SUCCESS)) {
    return <Redirect to={AppPath.NOT_FOUND} />;
  }

  useEffect(
      () => {
        onLoadOffer(activeOfferId);
        onLoadNearOffers(activeOfferId);
        onLoadReviews(activeOfferId);
      },
      [activeOfferId]
  );

  const handelSubmitReview = (review) => {
    onAddReview(activeOfferId, review);
  };

  const args = extend(props, {
    offer,
    handelSubmitReview,
  });

  const offerContent = offerStatus === DataStatus.SUCCESS ?
    getOfferContent(args) :
    null;

  return (
    <PageContainer type={ContainerType.PAGE}>
      <LoadingData status={offerStatus} />
      {offerContent}
    </PageContainer>
  );
};

OfferPage.propTypes = {
  offerStatus: Type.DATA_STATUS,
  onGetOffer: Type.FUNCTION,
  nearOffers: Type.LIST_OFFERS,
  nearOffersStatus: Type.DATA_STATUS,
  reviewsStatus: Type.DATA_STATUS,
  reviews: Type.REVIEWS,
  match: Type.MATCH,
  onLoadOffer: Type.FUNCTION,
  onLoadReviews: Type.FUNCTION,
  onLoadNearOffers: Type.FUNCTION,
  onAddReview: Type.FUNCTION,
  userStatus: Type.USER_STATUS,
  user: Type.USER,
};


const mapStateToProps = (state) => ({
  offerStatus: OffersSelector.getOfferStatus(state),
  onGetOffer: (id) => OffersSelector.getOffer(state, id),
  nearOffers: OffersSelector.getNearOffers(state),
  nearOffersStatus: OffersSelector.getNearOffersStatus(state),
  reviewsStatus: OffersSelector.getOfferReviewsStatus(state),
  reviews: OffersSelector.getOfferReviews(state),
  userStatus: UserSelector.getUserStatus(state),
  user: UserSelector.getUser(state),
});

const mapDispatchToPorps = (dispatch) => ({
  onLoadOffer: (id) => {
    dispatch(OffersOperation.loadOfferAsync(id));
  },
  onLoadReviews: (id) => {
    dispatch(OffersOperation.loadOfferReviewsAsync(id));
  },
  onLoadNearOffers: (id) => {
    dispatch(OffersOperation.loadNearOffersAsync(id));
  },
  onAddReview: (id, review) => {
    dispatch(OffersOperation.addOfferReviewAsync(id, review));
  },
});


export {OfferPage};
export default connect(mapStateToProps, mapDispatchToPorps)(OfferPage);
