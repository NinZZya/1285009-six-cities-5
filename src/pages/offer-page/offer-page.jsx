import React, {PureComponent} from 'react';
import {Redirect} from 'react-router-dom';
import {connect} from 'react-redux';
import PageContainer from '../../components/page-container/page-container';
import Container from '../../components/container/container';
import OfferGallery from '../../components/offer-gallery/offer-gallery';
import OfferMark from '../../components/offer-mark/offer-mark';
import BookmarkButton, {BookmarkButtonType} from '../../components/bookmark-button/bookmark-button';
import OfferFeatures from '../../components/offer-features/offer-features';
import OfferFeature from '../../components/offer-features/components/offer-feature/offer-feature';
import OfferPrice from '../../components/offer-price/offer-price';
import OfferInside from '../../components/offer-inside/offer-inside';
import OfferHost from '../../components/offer-host/offer-host';
import Reviews from '../../components/reviews/reviews';
import NewReview from '../../components/reviews/components/new-review/new-review';
import Map from '../../components/map/map';
import LoadingData from '../../components/loading-data/loading-data';
import OffersList, {OffersListType} from '../../components/offers-list/offers-list';
import * as UserSelector from '../../reducer/user/user-selectors';
import * as OffersOperation from '../../reducer/offers/offers-operations';
import * as OffersSelector from '../../reducer/offers/offers-selectors';
import * as Type from '../../types';
import {AppPath, IdName, DataStatus, UserStatus} from '../../const';
import RaitingStars from '../../components/raiting-stars/raiting-stars';


const NEAR_OFFERS_COUNT = 3;
const ContainerType = {
  PAGE: `property`,
  GALLERY: `property__gallery`,
};

const FetureType = {
  ENTIRE: `entire`,
  BEDROOMS: `bedrooms`,
  ADULTS: `adults`,
};

const TypeName = {
  OFFER_PRICE: `property`,
  RAITING_STARS: `property__stars`,
  MARK: `property`,
};

const OFFER_CONTAINER_CLASS_NAME = `property__container`;

const getOfferId = (match) => {
  const matchId = match.params[IdName.OFFER];
  const isCoorectId = matchId ? matchId.slice(0, 2) !== `00` : false;
  const offerId = Number(matchId);

  return !isNaN(offerId) && isCoorectId ? offerId : -1;
};

class OfferPage extends PureComponent {
  constructor(props) {
    super(props);

    this._activeOfferId = -1;
    this._offer = null;
    this._handelSubmitReview = this._handelSubmitReview.bind(this);
  }

  componentDidMount() {
    const {
      loadOffer,
      loadReviews,
      loadNearOffers,
    } = this.props;

    loadOffer(this._activeOfferId);
    loadNearOffers(this._activeOfferId);
    loadReviews(this._activeOfferId);
  }

  _handelSubmitReview(review) {
    const {
      addReview,
    } = this.props;

    addReview(this._activeOfferId, review);
  }

  _renderReviewCount(count) {
    return (
      <>
        &middot; <span className="reviews__amount">{count}</span>
      </>
    );
  }

  _getReviewsContent() {
    const {
      userStatus,
      user,
      reviewsStatus,
      reviews,
    } = this.props;

    const count = reviews.length;
    const isAuth = userStatus === UserStatus.AUTH && user;

    const reviewsContent = (
      reviewsStatus !== DataStatus.LOADING && reviewsStatus !== DataStatus.ERROR
    ) ?
      <Reviews reviews={reviews} /> :
      null;

    return (
      <section className="property__reviews reviews">
        <h2 className="reviews__title">
          Reviews {count ? this._renderReviewCount(count) : null}
        </h2>
        <LoadingData status={reviewsStatus} />
        {reviewsContent}
        {isAuth && (
          <NewReview
            reviewsStatus={reviewsStatus}
            onSubmitReview={this._handelSubmitReview}
          />
        )}
      </section>
    );
  }

  _getOfferContent() {
    const {
      nearOffersStatus,
    } = this.props;

    const {
      images,
      city,
      isPremium,
      isFavorite,
      title,
      rate,
      type,
      bedroomsCount,
      adultsCount,
      features,
      price,
      host,
      description,
      coords,
    } = this._offer;

    const nearOffers = this.props.nearOffers.slice(0, NEAR_OFFERS_COUNT);
    const mapNearOffers = [...nearOffers, this._offer];
    const nearOffersContent = nearOffersStatus === DataStatus.SUCCESS ?
      <OffersList type={OffersListType.NEAR} offers={nearOffers} /> :
      null;

    const center = {
      coords,
      zoom: city.zoom,
    };

    return (
      <>
        <section className="property">
          <Container type={ContainerType.GALLERY}>
            <OfferGallery images={images} />
          </Container>
          <Container className={OFFER_CONTAINER_CLASS_NAME}>
            <div className="property__wrapper">
              {isPremium && <OfferMark type={TypeName.MARK} />}
              <div className="property__name-wrapper">
                <h1 className="property__name">
                  {title}
                </h1>
                <BookmarkButton
                  type={BookmarkButtonType.OFFER}
                  mark={isFavorite}
                  offer={this._offer}
                />
              </div>
              <div className="property__rating rating">
                <RaitingStars type={TypeName.RAITING_STARS} rate={rate} />
                <span className="property__rating-value rating__value">{rate}</span>
              </div>
              <OfferFeatures>
                <OfferFeature type={FetureType.ENTIRE} value={type} />
                <OfferFeature type={FetureType.BEDROOMS} value={`${bedroomsCount} Bedrooms`} />
                <OfferFeature type={FetureType.ADULTS} value={`Max ${adultsCount} adults`} />
              </OfferFeatures>
              <OfferPrice type={TypeName.OFFER_PRICE} price={price} />
              {features.length ? <OfferInside features={features} /> : null}
              <OfferHost host={host} description={description} />
              {this._getReviewsContent()}
            </div>
          </Container>
          <section className="property__map map">
            <Map
              center={center}
              pins={mapNearOffers}
              activeId={this._offer.id}
            />
          </section>
        </section>
        <Container>
          <section className="near-places places">
            <h2 className="near-places__title">
              Other places in the neighbourhood
            </h2>
            <LoadingData status={nearOffersStatus} />
            {nearOffersContent}
          </section>
        </Container>
      </>
    );
  }

  render() {
    const {
      offerStatus,
      getOffer,
      match,
    } = this.props;

    this._activeOfferId = getOfferId(match);
    this._offer = offerStatus === DataStatus.SUCCESS ?
      getOffer(this._activeOfferId) :
      null;

    if (this._activeOfferId === -1 || (!this._offer && offerStatus === DataStatus.SUCCESS)) {
      return <Redirect to={AppPath.NOT_FOUND} />;
    }

    const offerContent = offerStatus === DataStatus.SUCCESS ?
      this._getOfferContent() :
      null;

    return (
      <PageContainer type={ContainerType.PAGE}>
        <LoadingData status={offerStatus} />
        {offerContent}
      </PageContainer>
    );
  }
}

OfferPage.propTypes = {
  offerStatus: Type.DATA_STATUS,
  getOffer: Type.FUNCTION,
  nearOffers: Type.LIST_OFFERS,
  nearOffersStatus: Type.DATA_STATUS,
  reviewsStatus: Type.DATA_STATUS,
  reviews: Type.REVIEWS,
  match: Type.MATCH,
  loadOffer: Type.FUNCTION,
  loadReviews: Type.FUNCTION,
  loadNearOffers: Type.FUNCTION,
  addReview: Type.FUNCTION,
  userStatus: Type.USER_STATUS,
  user: Type.USER,
};


const mapStateToProps = (state) => ({
  offerStatus: OffersSelector.getOfferStatus(state),
  getOffer: (id) => OffersSelector.getOffer(state, id),
  nearOffers: OffersSelector.getNearOffers(state),
  nearOffersStatus: OffersSelector.getNearOffersStatus(state),
  reviewsStatus: OffersSelector.getOfferReviewsStatus(state),
  reviews: OffersSelector.getOfferReviews(state),
  userStatus: UserSelector.getUserStatus(state),
  user: UserSelector.getUser(state),
});

const mapDispatchToPorps = (dispatch) => ({
  loadOffer: (id) => {
    dispatch(OffersOperation.loadOfferAsync(id));
  },
  loadReviews: (id) => {
    dispatch(OffersOperation.loadOfferReviewsAsync(id));
  },
  loadNearOffers: (id) => {
    dispatch(OffersOperation.loadNearOffersAsync(id));
  },
  addReview: (id, review) => {
    dispatch(OffersOperation.addOfferReviewAsync(id, review));
  },
});


export {OfferPage};
export default connect(mapStateToProps, mapDispatchToPorps)(OfferPage);
