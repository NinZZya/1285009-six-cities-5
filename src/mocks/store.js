import NameSpace from '../store/reducer/name-space';
import offers, {offersMap} from './offers';
import cities from './cities';
import reviews from './reviews';
import {user} from './users';
import {SortType, DataStatus, UserStatus} from '../constants/const';

const ACTIVE_CITY_ID = 4;
const ERROR = `ERROR`;
const DATA_STATUS = DataStatus.SUCCESS;
const SORT_TYPE = SortType.HIGH_TO_LOW;

const citiesState = {
  cities,
  activeId: ACTIVE_CITY_ID,
};

const offersState = {
  status: DATA_STATUS,
  offers: offersMap,
  favorites: offersMap,
  sortType: SORT_TYPE,
  offerStatus: DATA_STATUS,
  reviewsStatus: DATA_STATUS,
  reviews,
  nearOffersStatus: DATA_STATUS,
  nearOffers: offers,
};

const userState = {
  status: UserStatus.NO_AUTH,
  user,
  error: ERROR,
};

export default {
  [NameSpace.CITIES]: citiesState,
  [NameSpace.OFFERS]: offersState,
  [NameSpace.USER]: userState,
};
