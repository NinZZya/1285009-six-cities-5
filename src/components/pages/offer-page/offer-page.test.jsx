import React from 'react';
import {Router} from 'react-router-dom';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import {OfferPage} from './offer-page';
import store from '../../../store/store';
import history from '../../../constants/history';
import {DataStatus, UserStatus} from '../../../constants/const';
import offers from '../../../mocks/offers';
import reviews from '../../../mocks/reviews';
import {user} from '../../../mocks/users';


const match = {
  path: `offer/1`,
  params: {
    offerId: ``,
  },
};

const testing = () => {};
const getOffer = (id) => offers[id];


describe(`Should OfferPage render correctly`, () => {
  it(`Should OfferPage render correctly loading and nearOffers loading, and comments loading`, () => {
    const tree = renderer
      .create(
          <Router history={history}>
            <OfferPage
              offersStatus={DataStatus.LOADING}
              getOffer={getOffer}
              nearOffersStatus={DataStatus.LOADING}
              nearOffers={[]}
              reviewsStatus={DataStatus.LOADING}
              reviews={[]}
              match={match}
              loadOffer={testing}
              loadReviews={testing}
              loadNearOffers={testing}
              addReview={testing}
              userStatus={UserStatus.NO_AUTH}
              user={user}
            />
          </Router>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly sucsess and nearOffers loading, and comments loading`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.SUCCESS}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.LOADING}
                nearOffers={offers}
                reviewsStatus={DataStatus.LOADING}
                reviews={[]}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.NO_AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly sucsess and nearOffers sucsess, and comments loading`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.SUCCESS}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.SUCCESS}
                nearOffers={offers}
                reviewsStatus={DataStatus.LOADING}
                reviews={reviews}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.NO_AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly sucsess and nearOffers sucsess, and comments sucsess`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.SUCCESS}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.SUCCESS}
                nearOffers={[]}
                reviewsStatus={DataStatus.SUCCESS}
                reviews={reviews}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.NO_AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly sucsess and nearOffers sucsess, and comments sucsess, and user auth`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.SUCCESS}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.SUCCESS}
                nearOffers={[]}
                reviewsStatus={DataStatus.SUCCESS}
                reviews={reviews}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly error`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.ERROR}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.LOADING}
                nearOffers={[]}
                reviewsStatus={DataStatus.LOADING}
                reviews={[]}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.NO_AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly sucsess and nearOffers error, and comments loading`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.SUCCESS}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.ERROR}
                nearOffers={[]}
                reviewsStatus={DataStatus.LOADING}
                reviews={reviews}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should OfferPage render correctly sucsess and nearOffers sucsess, and comments error`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                offersStatus={DataStatus.SUCCESS}
                getOffer={getOffer}
                nearOffersStatus={DataStatus.SUCCESS}
                nearOffers={[]}
                reviewsStatus={DataStatus.ERROR}
                reviews={reviews}
                match={match}
                loadOffer={testing}
                loadReviews={testing}
                loadNearOffers={testing}
                addReview={testing}
                userStatus={UserStatus.AUTH}
                user={user}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
