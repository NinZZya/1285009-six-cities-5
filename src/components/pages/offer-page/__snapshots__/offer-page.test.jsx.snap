// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should OfferPage render correctly Should OfferPage render correctly error 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly loading and nearOffers loading, and comments loading 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly sucsess and nearOffers error, and comments loading 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly sucsess and nearOffers loading, and comments loading 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly sucsess and nearOffers sucsess, and comments error 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly sucsess and nearOffers sucsess, and comments loading 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly sucsess and nearOffers sucsess, and comments sucsess 1`] = `null`;

exports[`Should OfferPage render correctly Should OfferPage render correctly sucsess and nearOffers sucsess, and comments sucsess, and user auth 1`] = `null`;
